docs/zkapps/tutorials/04-in-browser-react-ui.mdx

```diff
@@ -0,0 +1,407 @@
+
+### What we will build
+
+The application we will build will implement the following:
+1. Loading a PrivateKey from the page's localstorage if one exists, otherwise create one.
+2. Checking if the private key has funds, and direct the user to the faucet if not.
+3. Connecting to example zkapp `Add` smart contract, already deployed on berkeley at a fixed address.
+4. A button that when clicked sends a transaction.
+5. A button that when clicked requests the latest state of the smart contract.
+
+Like in tutorial 3, we will provide a couple helper files, so we can focus on the react implementation itself. What is special about these helpers though, is that they use a webworker. This ensures the UI thread isn't blocked during long computations like compiling a smart contract or proving a transaction.
+
+This example uses an RPC endpoint. An in-browser Mina node is in the works, which will provide full node level security to your users, in the browser.
+
+While this current tutorial uses localStorage to store and manage the user's private key, a new version of SnarkyJS is in the works which will connect to Auro wallet for more persistent and secure private key management.
+
+## Project setup
+
+First, setup a new project with
+
+```sh
+zk project 04-zkapp-browser-ui --ui next
+```
+
+On the prompts that appear, select:
+1. Do you want your NextJS project to use TypeScript? `yes`
+2. Do you want to setup your project for github pages? `yes`
+
+This will create a new project with two folders instead of just one:
+* `contract`: This stores your smart contract code
+* `ui`: This is where you will write your ui
+
+We will be using the default contract (`Add`), so all we'll be doing is building it for use from our ui.
+
+To do this, enter the contract folder, and run 
+```sh
+npm run build
+```
+
+In making your own zkApp, you will need to edit the contract folder, and rebuild before accessing it from your UI.
+
+## Implementing the UI
+
+Our React UI will have a few components. First, there is the react page itself. In addition though, there is code that uses SnarkyJS.
+
+Because SnarkyJS code is computationally intensive, running it as usual in a script would block our browser's UI thread, causing our page to become unresponsive at times. To solve this, we will put our SnarkyJS code in a webworkers. 
+
+This will be implemented via 2 helper files:
+* `zkappWorker.ts`: The web worker code itself
+* `zkappWorkerClient.ts`: Code that we run from react to interact with the web worker.
+
+We encourage you to read this files, to understand how they work and to extend them for your own zkApp, but we will not review them in detail here, in favor of spending time on the react code.
```

Comment:
This tutorial makes it feel dead easy to use web workers with snarkyjs 💯 

I wondered at the beginning whether we should show an advanced pattern like workers in our very first tutorial with a UI. In the end, I think it's worth it -- worth making the architecture feel a bit more complex to users, for the gain that users will use the right architecture from the beginning, and that snarkyjs UIs will feel way less clunky. So, I'm +1 on this. Also +1 on not showing the actual worker code here.